### PR TITLE
HPCC-17410 Add roxie cluster into WsDFUXRef.DFUXRefList

### PR DESCRIPTION
**Pull request by wangkx**
Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by wangkx** [comment]
@afishbeck please review.

File: esp/services/ws_dfu/ws_dfuXRefService.cpp [hunk 1]
```diff
@@ -599,17 +611,19 @@ bool CWsDfuXRefEx::onDFUXRefList(IEspContext &context, IEspDFUXRefListRequest &r
...
-                const char *thorProcess = primaryThorProcesses.item(i);
-                if (!thorProcess || !*thorProcess)
-                    continue;
-                bool* found = uniqueProcesses.getValue(thorProcess);
-                if (found && *found)
-                    continue;
-                uniqueProcesses.setValue(thorProcess, true);
-                addXRefNode(thorProcess, pXRefNodeTree);
+                if (addUniqueXRefNode(primaryThorProcesses.item(i), uniqueProcesses, pXRefNodeTree))
+                    isThorCluster = true;
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuXRefService.cpp]: I find setting this flag and checking it a little less clear than just wrapping this section in:

`switch (cluster.getPlatform())`
`{`
`case ThorLCRCluster:`
`case RoxieCluster:`
`}`

**Review by afishbeck — CHANGES_REQUESTED** [review]

**Comment by afishbeck** [comment]
@wangkx back to you

**Comment by wangkx** [comment]
@afishbeck I made the changes. Please review.

File: esp/services/ws_dfu/ws_dfuXRefService.cpp [hunk 2]
```diff
@@ -599,17 +611,19 @@ bool CWsDfuXRefEx::onDFUXRefList(IEspContext &context, IEspDFUXRefListRequest &r
...
-                addXRefNode(thorProcess, pXRefNodeTree);
+            case ThorLCRCluster:
+                {
+                    const StringArray &primaryThorProcesses = cluster.getPrimaryThorProcesses();
+                    ForEachItemIn(i, primaryThorProcesses)
+                        addUniqueXRefNode(primaryThorProcesses.item(i), uniqueProcesses, pXRefNodeTree);
+                }
+                break;
+            case RoxieCluster:
+                SCMStringBuffer roxieProcess;
```
Comment (afishbeck) [inline esp/services/ws_dfu/ws_dfuXRefService.cpp]: Trivial, but I usually put braces {} around any "case" with local variables.. even though in some cases like this one it's probably not necessary.

**Review by afishbeck — APPROVED** [review]
Trivial comment, but looks fine.

**Pull request merged**